submissions/DmitryHniezdilov/task_js_oop/app.js

```diff
@@ -0,0 +1,349 @@
+"use strict";
+
+const initialInterfaceMarkup = () => {
+  document.body.innerHTML = `
+    <h1 class='title'>Frogger Game</h1>
+    <div class='info-wrap'>
+      <p><span>Level: </span><span class='js-info-level'>1</span></p>
+      <p><span>Speed: </span><span class='js-info-speed'>x1</span></p>
+    </div>
+    <p class='popup js-popup'></p>
+  `;
+};
+
+initialInterfaceMarkup();
+
+const allEnemies = [];
+
+const config = {
+  BLOCK_WIDTH: 101,
+  BLOCK_HEIGHT: 83,
+  BLOCKS_NUMBER: {
+    x: 5,
+    y: 6,
+  },
+  get FIELD_WIDTH() {
+    return this.BLOCK_WIDTH * this.BLOCKS_NUMBER.x;
+  },
+  ORIGIN_COORDINATE_CHARACTERS: {
+    x: 0,
+    y: -20,
+  },
+  PLAYER_CONF: {
+    img: "images/char-boy.png",
+    initialBlock: {
+      x: 3,
+      y: 6,
+    },
+  },
+  ENEMIES_CONF: {
+    img: "images/enemy-bug.png",
+    initialRows: [2, 3, 4],
+    amountOfEnemies: [2, 2, 1],
+    overlapRatio: 1.3,
+    speedIncrease: 0.2,
+    minSpeedOfEnemy: 50,
+    maxSpeedOfEnemy: 300,
+  },
+  STATE: {
+    currentLevel: 1,
+    speedMultiplicator: 1,
+  },
+  END_OF_GAME_WIN: "win",
+  END_OF_GAME_LOSE: "lose",
+  PRINT_MESSAGE_WIN: "You Win",
+  PRINT_MESSAGE_LOSE: "You Lose",
+  ELEMENT_POPUP_MESSAGE: document.querySelector(".js-popup"),
+  ELEMENT_INFO_LEVEL: document.querySelector(".js-info-level"),
+  ELEMENT_INFO_SPEED: document.querySelector(".js-info-speed"),
+  POPUP_MESSAGE_ACTIVE_CLASS: "is-active",
+};
+
+const Character = function (imgUrl) {
+  this.sprite = imgUrl;
+  this.isFreeze = false;
+};
+
+Character.prototype.setFreeze = function () {
+  this.isFreeze = true;
+};
+
+Character.prototype.unFreeze = function () {
+  this.isFreeze = false;
+};
+
+const Enemy = function (x, y, speed, config, endOfGame) {
+  Character.call(this, config.ENEMIES_CONF.img);
+  this.config = config;
+  this.STATE = config.STATE;
+  this.ENEMIES_CONF = config.ENEMIES_CONF;
+  this.FIELD_WIDTH = config.FIELD_WIDTH;
+  this.BLOCK_WIDTH = config.BLOCK_WIDTH;
+  this.BLOCK_HEIGHT = config.BLOCK_HEIGHT;
+  this.END_OF_GAME_LOSE = config.END_OF_GAME_LOSE;
+  this.x = x;
+  this.y = y;
+  this.speed = speed;
+  this.endOfGame = endOfGame;
+};
+
+Enemy.prototype = Object.create(Character.prototype);
+
+Enemy.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Enemy.prototype.update = function (dt) {
+  !this.isFreeze &&
+    ((this.x += this.speed * dt),
+    this.x > this.FIELD_WIDTH && (this.x = -this.BLOCK_WIDTH),
+    this.checkCollision());
+};
+
+Enemy.prototype.checkCollision = function () {
+  const isCollision =
+    player.y === this.y &&
+    player.x <=
+      Math.floor(this.x) + this.BLOCK_WIDTH / this.ENEMIES_CONF.overlapRatio &&
+    player.x >=
+      Math.floor(this.x) - this.BLOCK_WIDTH / this.ENEMIES_CONF.overlapRatio;
+
```

Comment:
See item 4.iii of the [requirements](https://kottans.org/documentation/docs/doc/code-review/#oo-js-frogger)